dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp

```diff
@@ -123,6 +156,40 @@ void DistributedBlockOutputStream::writeToLocal(const Block & block, const size_
 }
 
 
+void DistributedBlockOutputStream::writeToShardDirect(const Block & block, const std::vector<std::string> & dir_names, std::atomic<bool> & timeout_exceeded)
+{
+    const auto & query_string = queryToString(query_ast);
+    for (const auto & dir_name : dir_names)
+    {
+        auto & monitor = storage.requireDirectoryMonitor(dir_name);
+        auto & pool = monitor.getPool();
+        auto connection = pool->get();
+
+        CurrentMetrics::Increment metric_increment{CurrentMetrics::DistributedSend};
+
+        if (timeout_exceeded)
+            throw Exception("Timeout exceeded. Inserted blocks: " + std::to_string(blocks_inserted), ErrorCodes::TIMEOUT_EXCEEDED);
```

Comment:
Probably there will be many similar messages for each shard. But these messages doesn't contain shard name and indistinguishable.